runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -3916,6 +3917,131 @@ J9::CodeGenerator::collectSymRefs(
+      TR::Node *node = treeTop->getNode();
+      if (node->getOpCode().isIf() && node->isTheVirtualGuardForAGuardedInlinedCall() && !checklist.contains(node))
+         {
+         TR_VirtualGuard *vg = comp->findVirtualGuardInfo(node);
+         // Mainly we need to make sure that virtual guard which performs the TR_MethodTest and can be NOP'd are needed the range check.
+         if (vg && vg->getTestType() == TR_MethodTest && 
+            !(comp->performVirtualGuardNOPing() && (node->isNopableInlineGuard() || comp->isVirtualGuardNOPingRequired(vg))))
+            {
+            TR::SymbolReference *callSymRef = vg->getSymbolReference();
+            TR_ASSERT_FATAL(callSymRef != NULL, "Guard n%dn for the inlined call should have stored symbol reference for the call", node->getGlobalIndex());
+            if (callSymRef->getSymbol()->castToMethodSymbol()->isInterface())
+               {
+               TR::DebugCounter::incStaticDebugCounter(comp, TR::DebugCounter::debugCounterName(comp, "profiledInterfaceTest/({%s}{%s})", comp->signature(), comp->getHotnessName(comp->getMethodHotness())));
+               dumpOptDetails(comp, "Need to add a rangecheck before n%dn in block_%d\n",node->getGlobalIndex(), block->getNumber());
+
+               // We need a VFT Load of the receiver object to get the VTableHeader.size to check the range. As this operation is happening during codegen phase, only
+               // known concrete way we can have this information is through aloadi child of the guard that has single child which is vft load of receiver object.
+               // Now instead of accessing VFT load from the child of the aloadi, we could have treetop's the aloadi during inlining where we generate the virtual guard
+               // to access information from the treetop. Because of the same reasons lined up behind adding range check test during codegen phase in the description of this function,
+               // we would need to make changes in all optimizations moving Virtual Guard around to keep that treetop together before guard which will be very difficult to enforce.
+               // Also as children of virtual guard is very self contained and atm it is very unlikely that other optimizations are going to find opportunity of manipulating them and
+               // Because of the fact that it is very unlikely that we will have another aloadi node with same VTable offset of same receiver object, this child would not be commoned out
+               // and have only single reference in this virtual guard therefore splitting of block will not store it to temp slot.
+               // Due to the above reasons we have a fatal assert in case the first child of the guard node performing is not aloadi. If there exists any case we need to reconsider our choice for range check test.
+               TR::Node *vTableLoad = node->getFirstChild();
+               TR_ASSERT_FATAL(vTableLoad->getOpCodeValue() == TR::aloadi &&
```

Comment:
Given that Devin has seen the aloadi happen I don't think we can tolerate a fatal assert in production - would you crash the JVM? I think the right thing is to abort the compilation... It should be rare but a fatal assert when we have a known example makes me worry that we would generate issues for users needlessly

Reply:
Another alternative to failing the compile is to do a type test or to jump unconiditionally to the failed side of the guard - the latter probably being the easiest.

Reply:
I am getting inclined towards the idea jump unconditionally to prevent JVM for crashing. This would definitely have lesser perf impact then failing compilation but I have one concern with that. In case, some one saves `vtable-entry-table` symbol in temp slot (hoisting out of the loop or splitting block, etc), that means we would still have a bug present in the compiled body where illegal reference appears well before of the guard. Skipping over the whole guard and inlined body in that case will not prevent the illegal memory reference.

Reply:
> Another alternative to failing the compile is to do a type test or to jump unconiditionally to the failed side of the guard - the latter probably being the easiest.

I agree. The safe thing to do is always bail to the slow path. We can place a static debug counter here to keep track of how often we need fall back to the slow path if we want such data (in customer scenarios). 

@andrewcraik do we have some sort of debug counter name prepend which is meant for "bad performance" things that we're doing?

Reply:
not at present - we were going to look to do that as part of the assert conversion work - we hadn't agreed a convention yet.

Reply:
I haven't actually seen `aloadi <vtable-entry-symbol>` get replaced with a load of a temp, but I have seen the corresponding thing happen to `aloadi <vft-symbol>`, and I know of a couple of situations in which it would be reasonable to do that transformation with `<vtable-entry-symbol>` as well, unless we decide that it's illegal to do so.

----

The first situation is as follows. We have two calls, one after the other, calling the same method of the same object.
```
treetop
  calli Foo.bar()V
    aloadi <vft-symbol>
      aload obj
    ==>aload obj
treetop
  calli Foo.bar()V
    aloadi <vft-symbol>
      aload obj
    ==>aload obj
```
We inline both with the same profiled guard with method test, and virtual guard tail splitter causes the second guard to be dominated by the first inlined body. After block extension, local CSE commons the children of the guards:
```
BBStart
ifacmpne --> slow call 0
  aloadi <vtable-entry-symbol>
    aloadi <vft-symbol>
      aload obj
  aconst 0x...
BBEnd
BBStart (extension)
<...inlined Foo.bar()V 0...>
BBEnd
BBStart (extension)
ifacmpne --> slow call 1
  ==>aloadi <vtable-entry-symbol>
  ==>aconst 0x...
BBEnd
BBStart (extension)
<...inlined Foo.bar()V 1...>
BBEnd
```
If `<...inlined Foo.bar()V 0...>` is split and a merge point is inserted, then `aloadi <vtable-entry-symbol>` will be spilled into a temp.

Assuming this kind of thing is sufficiently unlikely, I believe it's fine to deal with it by unconditionally failing the guard, as mentioned upthread.

----

The second scenario is code motion, as @r30shah mentioned. To elaborate a bit, the safety requirements for `aloadi <vtable-entry-symbol>` are stricter than those for `<vft-symbol>`, and I don't know whether anything checks for those and moves these loads. However, it would be reasonable to hoist any (invariant) load whatsoever if, upon entering the loop, the load would necessarily run, e.g.
```
do {
    x = obj.baz;  // NB. if loop has been entered, this *always* runs
    ...
} while(...);
```
can in general be safely rewritten as
```
tmp = obj.baz;
do {
    x = tmp;
    ...
} while(...);
```
because if the load is disallowed at the new location, it would also have been disallowed at its original location, no matter what its safety conditions are.

The issue would be that with this change, when `aloadi <vtable-entry-symbol>` occurs in a profiled guard, the vtable entry isn't necessarily loaded every time we reach the guard. It has an implicit safety condition that is tested at runtime. Because the test is implicit, it might look as though the load can be hoisted, but hoisting it would remove it from the guard, and as a result separate it from the implicit test.

To ensure that we don't hit this kind of problem in the future, I think we have to do one of the following:
1. decree that no optimization may speculatively load `<vtable-entry-symbol>`, at least when its original occurrence was under a profiled guard; or
2. decree that no optimization may speculatively load `<vtable-entry-symbol>` without first checking the size of the vtable, _even if_ the existing control flow appears to guarantee that the vtable is large enough; or
3. create a new node that encapsulates the conditional, which would be safe to speculatively evaluate; or
4. make the safety test explicit in the CFG.

Personally I'm not a fan of restrictions like 1 or 2 (out of which 2 seems less objectionable to me), but I understand if given the alternatives one of them seems to be the best option, or at least the best option for the moment.

If we do want to go with something like 1 or 2, I'm not suggesting that we have to immediately audit all optimizations that could insert problematic speculative loads. This is more of a request to consider and possibly document the direction of the design w.r.t. this sort of issue

Reply:
Considering everything, It seems like jumping to slower path would not prevent the bug.
As we are checking both opcode of the child as well as type of symbol reference, If we do not have both, it would be safer to fail compilation. 
I have changed the code in https://github.com/eclipse/openj9/pull/4721/commits/0f44e093a81b6d6cb75ce55ccfef8dd1033e9a69